### DISCUSSION TITLE
Synchronization between management nodes

### DISCUSSION THREAD

**Discussion by oxedions** [bluebanquise/bluebanquise#604] [Ideas]
Hello everyone, 🙂 

In order to converge to BB 1.5, I need to solve one last main issue: how 
to synchronize data (`/etc/bluebanquise`, `/var/www/html/preboot_execution_environment/nodes`, `/var/www/html/repositories/`) 
between management nodes.

This is to be considered when High Availability but also with Multi Iceberg architecture. In the following, we consider "standard" usage, aka all managements are able to do everything (because one could choose to dedicate a management to repositories, another on to ansible, etc.)

* /etc/bluebanquise/ needs to be sync because all management nodes should be able to 
execute playbooks, and so need access to inventory, roles, playbooks, etc.
* /var/www/html/repositories/ needs to be sync because all management nodes should 
be able to provides repositories.
* /var/www/html/preboot_execution_environment/nodes is only when High Availability.
bootset tool need to be able to be used in HA environment, and so using bootset 
on one of the HA members should be seen by others. (especially if using ha_proxy role).

I explored what could be done, and found the following solutions:

1. Use NFS. Using NFS (in HA, one main resource, and clone resources for the others on client side) works
well, and is very simple. Main drawback are:
  * needs good network
  * kind of "freeze" nodes when servers is down (df is hanging, etc). Can be solved using soft mount, but dangerous
  * in case of nfs server crash, managements servers cannot provides some resources anymore, so big SPOF

2. Rsync. Need manual operation, or using a cron.
I found dangerous since you will probably need the --delete option to prevent trash. 
This is a simple solution, that could be documented.
So main drawback are:
  * kind of manual operation triggering
  * dangerous if using the --delete option

3. Lsyncd. This is an interesting solution, as it is based on kernel io watch mechanism, so it is triggered each time a modification occur.
But Lsyncd seems unmaintained, and what happen if a management node is out of the pool during few minutes ?
It could be unsynchronized as Lsyncd only apply events, not the whole sync. This would be coupled with a regular rsync (cron, time service, etc.)
So:
  * Seems unmaintained
  * Need coupling with another solution to ensure long term sync

4. Syncthing. This tool was interesting, but it has a main drawback: it must run in user environment, for security reasons, and so is not able to sync ids, SELinux tags, etc.
So this could be ok for /etc/bluebanquise, but not for http files.

5. Using git + git lfs combined with a central repository, was thinking about Gitea (that can be running in HA if needed, or on an external small power system, etc).
I like this solution, as it solves few things:
  * All systems have a copy, so no SPOF in case of failure, production can continue
  * Versioning, so you can recover cluster state (ansible and repositories)
  * Kind of a central point where the whole cluster is "saved", so you can backup gitea /var/lib on a dedicated system, and you are able to redeploy more easily
  * Gitea provides PRs, Issues, Wiki, users management, etc, all included. Could be interesting for a team to follow life of a cluster, and track changes.
 
It has however drawbacks:
  * Need hooks to push sync on managements repositories (using git over ssh I guess)
  * More complex than a basic nfs

So I would go with the following:

Default is NFS, using hard mount for /etc/bluebanquise and repositories, and soft mount for preboot_execution_environment/nodes.
And documentation also cover the possibility of git based synchronization, with git based /etc/bluebanquise and git lfs based repositories, and nfs soft for preboot_execution_environment/nodes.

But I may have miss some things. Do you have any opinion on this matter ?
@johnnykeats @btravouillon @bouriquet @strus38 @ginomcevoy @thiagocardozo ?

**Comment by strus38** [comment]
I would go with minio ... and mount S3 bucket as FS...

**Comment by btravouillon** [comment]
It looks like you're looking for a single solution to different problems.

## /etc/bluebanquise

> /etc/bluebanquise/ needs to be sync because all management nodes should be able to
execute playbooks, and so need access to inventory, roles, playbooks, etc.

IMHO, playbooks and inventory should be stored within git repositories, so you can sync those easily across the management nodes. Most sites may already have some GitLab or Gitea instances, or even GitHub private accounts so it's no big deal.

An alternative to those services could be to setup a git daemon :
```
$ git clone --bare inventory /tmp/tmp.d7ss85xfGs/inventory.git
Cloning into bare repository '/tmp/tmp.d7ss85xfGs/inventory.git'...
done.
$ touch /tmp/tmp.d7ss85xfGs/inventory.git/git-daemon-export-ok
$ git daemon --reuseaddr --verbose --base-path=/tmp/tmp.d7ss85xfGs -- /tmp/tmp.d7ss85xfGs/inventory.git
```
```
$ git clone git://localhost/inventory.git /tmp/inventory
Cloning into '/tmp/inventory'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Total 3 (delta 0), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (3/3), done.
```
The roles should be distributed through ansible-galaxy or OS packages whenever possible. Custom roles should also be maintained in some git repositories, like inventories and playbooks.

**Comment by btravouillon** [comment]
## /var/www/html/repositories/

> /var/www/html/repositories/ needs to be sync because all management nodes should
be able to provides repositories.

If one really wants to clone the repositories across the management nodes, I would recommand [reposync](https://www.redhat.com/sysadmin/how-mirror-repository) (EL) or [rsyncmirror](https://help.ubuntu.com/community/Rsyncmirror) (Ubuntu).
Another solution would be to install a web server (can be HA) on the top iceberg and web proxy in other icebergs.

**Comment by btravouillon** [comment]
## /var/www/html/preboot_execution_environment/nodes

> /var/www/html/preboot_execution_environment/nodes is only when High Availability.
bootset tool need to be able to be used in HA environment, and so using bootset
on one of the HA members should be seen by others. (especially if using ha_proxy role).

If I remember correctly those are files modified by **bootset** then consumed during **pxeboot** over HTTPS. If so, this directory should be writeable on the management node where the sysadmins run the **bootset** command. For read access, a web proxy is a good alternative, I don't get the need to synchronize those files to multiple servers.

**Comment by oxedions** [comment]
Thanks a lot for your feedback!! 😊 

So:

* /etc/bluebanquise/{inventory,playbooks,roles/custom} to be put in a git. We all agree with that. I will document no git, basic git, bind to an external git, bind to a local gitea, and your git daemon commands (very nice ones, thanks !).

* /var/www/html/repositories/ is a bit more tricky, as I may have made a small error when designing the stack: repositories/* does not contain only repositories, but also "netboot" elements (kernels and initrd, and magic os repository for some distributions). So using reposync only will not be enough. I guess NFS or an rsync would be ok if correctly used. This would involve a manually trigger operation, but this is not a so bad idea (I do not very like fully automated unverbose things).

* /var/www/html/preboot_execution_environment/nodes this is the tricky one. In MI mode, no need to sync that one. But in HA, the problem is the following:
  1. If not using HAproxy, then issue is just that when you execute bootset tool, it is not HA aware, and so nodes/*.ipxe files are only modified on the current host. So if HA resource for pxe is on another host of the HA cluster, then the bootset will have no effect.
  2. If using HAproxy, not only you have the bootset issue, but also the cgi script for post autoinstall (the one that reset what bootset did).
So solution would be to find a way to sync this folder, or mount it with an NFS over all nodes (since you should not have high IO load neither modifications on the same file at the same time on multiple hosts), or upgrade bootset to an HTTP API tool, so that a deamon is running on each hosts (or on the pxe resource host if no HAproxy).

I also gave a look to minio. This tool is very interesting to provide object storage, I will consider it for the next version of BB as I was planning to work more on optional containerization.

**Reply by btravouillon** [reply]
> If not using HAproxy, then issue is just that when you execut bootset tool, it is not HA aware, and so nodes/*.ipxe files are only modified on the current host.

It should be possible to force usage of the command on the active node only. I've seen this on some storage appliances where one management node in the HA pair had the expected resources (the directory w/ `nodes/*.ipxe` here) and was the only node authorized to run some commands. The command on other node was printing something close to "You should run this command on primary management node". Couple this with a floating IP in HA for sysadmin to connect to the right node and you're good to go IMHO.

I'm not sure to understand the need of HAproxy and distributed web servers for those ipxe files. See https://www.nginx.com/blog/testing-the-performance-of-nginx-and-nginx-plus-web-servers/ for RPS numbers. With 2 CPUs and 10 KB files, nginx can reach circa 50k RPS. This would need additional testing with the CGI script, but I'm almost sure 99% of the use cases can fit in a single web server, with eventually proxies in front for distribution across islets.

**Reply by oxedions** [reply]
I see your point, so make use of PCS resources to only enable tool on one system. Maybe using links. That is an interesting move, I will test it, thanks for the tip !

Regarding HAproxy, well this is not for these files, neither the repositories to be honest.
Demand came for the diskless images, that could be a bit heavy, and so would require more bandwidth.

But HAproxy is a fully optional community role (not in the stack core, same for HA role), I am just thinking of a way to allow full compatibility. Maybe disklessset should allow optional handle of the floating ip for HAproxy, so that HAproxy is only used for diskless images (and if asked for, repositories also, using dedicated already available variables).
Nextserver is then kept untouched.
Would be the best I think, since current PXE process have proven very strong, and I would like to avoid modifying it.

All of this is and must remain optional, main stack does not change.
